spring-cloud-gcp-autoconfigure/src/main/java/org/springframework/cloud/gcp/autoconfigure/secretmanager/SecretManagerPropertySource.java

```diff
@@ -16,117 +16,119 @@
  */
 public class SecretManagerPropertySource extends EnumerablePropertySource<SecretManagerServiceClient> {
 
-	private static final Log LOGGER = LogFactory.getLog(SecretManagerPropertySource.class);
-
-	private static final String LATEST_VERSION_STRING = "latest";
-
-	private final Map<String, Object> properties;
+	private static final String GCP_SECRET_PREFIX = "gcp-secret/";
 
-	private final String[] propertyNames;
+	private final GcpProjectIdProvider projectIdProvider;
 
 	public SecretManagerPropertySource(
 			String propertySourceName,
 			SecretManagerServiceClient client,
-			GcpProjectIdProvider projectIdProvider,
-			String secretsPrefix) {
-		this(propertySourceName, client, projectIdProvider, secretsPrefix, Collections.EMPTY_MAP);
-	}
-
-	public SecretManagerPropertySource(
-			String propertySourceName,
-			SecretManagerServiceClient client,
-			GcpProjectIdProvider projectIdProvider,
-			String secretsPrefix,
-			Map<String, String> versions) {
+			GcpProjectIdProvider projectIdProvider) {
 		super(propertySourceName, client);
 
-		Map<String, Object> propertiesMap = createSecretsPropertiesMap(
-				client, projectIdProvider.getProjectId(), secretsPrefix, versions);
+		this.projectIdProvider = projectIdProvider;
+	}
 
-		this.properties = propertiesMap;
-		this.propertyNames = propertiesMap.keySet().toArray(new String[propertiesMap.size()]);
+	@Override
+	public Object getProperty(String name) {
+		SecretVersionName secretIdentifier = parseFromProperty(name, this.projectIdProvider);
+
+		if (secretIdentifier != null) {
+			return getSecretPayload(secretIdentifier);
+		}
+		else {
+			return null;
+		}
 	}
 
+	/**
+	 * The {@link SecretManagerPropertySource} is not enumerable, so this always returns an empty array.
+	 * @return the empty array.
+	 */
 	@Override
 	public String[] getPropertyNames() {
-		return propertyNames;
+		return new String[0];
 	}
 
-	@Override
-	public Object getProperty(String name) {
-		return properties.get(name);
+	private ByteString getSecretPayload(SecretVersionName secretIdentifier) {
+		try {
+			AccessSecretVersionResponse response = getSource().accessSecretVersion(secretIdentifier);
+			return response.getPayload().getData();
+		}
+		catch (NotFoundException e) {
+			return null;
+		}
 	}
 
-	private static Map<String, Object> createSecretsPropertiesMap(
-			SecretManagerServiceClient client, String projectId, String secretsPrefix, Map<String, String> versions) {
-
-		ListSecretsPagedResponse response = client.listSecrets(ProjectName.of(projectId));
-		Map<String, Object> secretsMap = new HashMap<>();
-		for (Secret secret : response.iterateAll()) {
-			String secretId = extractSecretId(secret);
-			ByteString secretPayload = getSecretPayload(client, projectId, secretId, versions);
-			if (secretPayload != null) {
-				secretsMap.put(secretsPrefix + secretId, secretPayload);
-			}
+	static SecretVersionName parseFromProperty(String property, GcpProjectIdProvider projectIdProvider) {
```

Comment:
Visible for testing?

Reply:
Added comment `//visible for testing`; can't add annotation though, it's guava.